repo2docker/buildpacks/base.py

```diff
@@ -414,12 +415,18 @@ def get_start_script(self):
         """
         return None
 
+    @property
+    def binder_dir(self):
+        if self._binder_dir is None:
+            for dirname in ['binder', '.binder']:
+                if os.path.exists(dirname):
+                    self._binder_dir = dirname
+            self._binder_dir = ''
```

Comment:
This is a problematic line right? You need this to be in a conditional to not finding any dirs

Reply:
Oh! Move it to before the for loop

Reply:
@betatim, is there a risk that this property is called before a binder / .binder folder exists, and then isn't updated to reflect that there now is a binder / .binder folder?

@jhamman I guess the fact that I don't know at this point, makes me think that we want to avoid the preoptimization only checking for the existence of the folders once.

Reply:
Not sure if we could end up with a race condition. Not caching this is a good way to side step the issue and I'd be surprised if this makes any noticeable difference to the speed of running repo2docker.

Reply:
I'd write something like this to make the intention of exclusive directories (very) explicit for future readers:
```python
@property
def binder_dir(self):
    has_binder = os.path.isdir("binder")
    has_dotbinder = os.path.isdir(".binder")

    if has_binder and has_dotbinder:
        raise RuntimeError("The repository contains both a 'binder' and a "
                           "'.binder' directory. However they are exclusive.")

    if has_dotbinder:
        return ".binder"
    elif has_binder:
        return "binder"
    else:
        return ""
```

Reply:
Thanks for the discussion here. I've updated to match @betatim's implementation.